### PR TITLE
[PW-799] Upgrade rails, ruby ang gems

### PR DESCRIPTION
**Pull request by florfrioli**
## Summary

- Upgrades
Ruby → upgrade to  2.7.4
Rails → upgrade to 6.1.4
Puma → upgrade to 5.5.2
Rubocop → upgrade to locked 0.89 version.
pg - devise - active admin - http party - active model serializer - sidekiq - sidekiq failures
- Added gems:
kaminari - pghero - versionist - vcr - testprof

- Added github actions config
- Hooks copied from MRP
- Updated readme
- Fixed annotate issue with rails 6.1 adding a route task

- [X] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [X] I have checked if my change needs adjustments in Active Admin
- [X] I have checked if my change needs adjustments in the API documentation
- [X] I have reviewed the files changed before releasing the pull request
- [X] I have updated the JIRA card associated with: column status, incurred hours and a comment for the QA team.
- [ ] BEFORE MERGE: I have updated the JIRA card column before merging the pull request

## Screenshots
N/A

## JIRA Card

[Link to the associated JIRA card.](https://widergy.atlassian.net/browse/PW-799)

### CODE REVIEW

File: .github/workflows/build-and-test.yml [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    # Set enviroment variables
+    env:
+      RAILS_ENV: test
+      PGHOST: 127.0.0.1
+      DATABASE_URL: postgres://rails-api-bootstrap@127.0.0.1:5432/rails-api-bootstrap_test
+      DB_USERNAME: rails-api-bootstrap
+      DB_PASSWORD: ""
+      REDIS_HOST: redis
+      AWS_REGION: ${{ secrets.AWS_REGION }}
+      AWS_REGION_PUSH: ${{ secrets.WIDERGYDEV_AWS_REGION_PUSH }}
```
Comment (icapuccio) [inline .github/workflows/build-and-test.yml]: Remove this because is not used anymore

File: .github/workflows/build-and-test.yml [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+    # Set enviroment variables
+    env:
+      RAILS_ENV: test
+      PGHOST: 127.0.0.1
+      DATABASE_URL: postgres://rails-api-bootstrap@127.0.0.1:5432/rails-api-bootstrap_test
+      DB_USERNAME: rails-api-bootstrap
+      DB_PASSWORD: ""
+      REDIS_HOST: redis
+      AWS_REGION: ${{ secrets.AWS_REGION }}
```
Comment (icapuccio) [inline .github/workflows/build-and-test.yml]: We can remove this too for now

File: .github/workflows/deploy-dev.yml [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+    # Set enviroment variables
+    env:
+      RAILS_ENV: test
+      PGHOST: 127.0.0.1
+      DATABASE_URL: postgres://rails-api-bootstrap@127.0.0.1:5432/rails-api-bootstrap_test
+      DB_USERNAME: rails-api-bootstrap
+      DB_PASSWORD: ""
+      REDIS_HOST: redis
+      AWS_REGION: ${{ secrets.AWS_REGION }}
```
Comment (icapuccio) [inline .github/workflows/deploy-dev.yml]: Same for this file

File: .github/workflows/build-and-test.yml [hunk 4]
```diff
@@ -0,0 +1,91 @@
+name: Build & Test
+
+on:
+  pull_request:
+    types: [opened, synchronize]
+  push:
+    branches:
+      - master
```
Comment (icapuccio) [inline .github/workflows/build-and-test.yml:8]: Check if we can add a regex here for release branches

File: .github/workflows/deploy-prod.yml [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+    # Set enviroment variables
+    env:
+      RAILS_ENV: test
+      PGHOST: 127.0.0.1
+      DATABASE_URL: postgres://rails-api-bootstrap@127.0.0.1:5432/rails-api-bootstrap_test
+      DB_USERNAME: rails-api-bootstrap
+      DB_PASSWORD: ""
+      REDIS_HOST: redis
+      AWS_REGION: ${{ secrets.AWS_REGION }}
```
Comment (icapuccio) [inline .github/workflows/deploy-prod.yml]: Same

File: .github/workflows/deploy-dev.yml [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+          sudo pip3 install awsebcli --upgrade
+      - name: Configure AWS Credentials
+        uses: aws-actions/configure-aws-credentials@v1
+        with:
+          aws-access-key-id: ${{ secrets.WIDERGYDEV_AWS_ACCESS_KEY_ID }}
+          aws-secret-access-key: ${{ secrets.WIDERGYDEV_AWS_SECRET_ACCESS_KEY }}
+          aws-region: us-east-1
+      - name: Deploy to AWS ElasticBeastalk
+        run:
+          eb deploy meter-reading-api-dev
```
Comment (icapuccio) [inline .github/workflows/deploy-dev.yml]: Check the environment names on these files

File: .platform/hooks/postdeploy/50_restart_sidekiq.sh [hunk 7]
```diff
@@ -1,7 +1,7 @@
 # #!/bin/bash
 
-# EB_APP_DEPLOY_DIR=$(/opt/elasticbeanstalk/bin/get-config platformconfig -k AppDeployDir)
 # EB_APP_USER=$(/opt/elasticbeanstalk/bin/get-config platformconfig -k AppUser)
+# EB_APP_DEPLOY_DIR="/var/app/current"
```
Comment (icapuccio) [inline .platform/hooks/postdeploy/50_restart_sidekiq.sh:4]: check also the swap.sh and mute_sidekiq.sh

File: Gemfile [hunk 8]
```diff
@@ -1,16 +1,16 @@
 source 'https://rubygems.org'
 git_source(:github) { |repo| "https://github.com/#{repo}.git" }
 
-ruby '2.7.1'
+ruby '2.7.4'
 
 # Bundle edge Rails instead: gem 'rails', github: 'rails/rails'
-gem 'rails', '~> 6.0'
+gem 'rails', '~> 6.1.4'
```
Comment (icapuccio) [inline Gemfile]: `~> 6.1`

File: Gemfile [hunk 9]
```diff
@@ -1,16 +1,16 @@
...
 
 # Use postgresql as the database for Active Record
 gem 'pg', '>= 0.18', '< 2.0'
 
 # Use Puma as the app server
-gem 'puma', '~> 4.3'
+gem 'puma', '~> 5.5.2'
```
Comment (icapuccio) [inline Gemfile]: `~> 5.5`

File: Gemfile [hunk 10]
```diff
@@ -29,11 +29,14 @@ gem 'activeadmin', '~> 2.3'
...
 
+# For PORO management
+gem 'interactor', '~> 3.0'
+
 # A serializer implementation for Ruby On Rails Objects.
 gem 'active_model_serializers', '~> 0.10.9'
 
 # Sidekiq
-gem 'sidekiq', '< 7'
+gem 'sidekiq', '<6'
```
Comment (icapuccio) [inline Gemfile]: Prefer having `< 7`

File: Gemfile [hunk 11]
```diff
@@ -128,20 +131,26 @@ gem 'bullet'
...
-  gem 'shoulda-matchers', '~> 4.0'
+  gem 'shoulda-matchers'
 
   # Code coverage
   gem 'simplecov', require: false
 
   # Mocking
+  gem 'timecop', '~> 0.9'
   gem 'webmock'
   # gem 'timecop', '~> 0.9'
```
Comment (icapuccio) [inline Gemfile]: Remove this line

File: Gemfile [hunk 12]
```diff
@@ -128,20 +131,26 @@ gem 'bullet'
...
-  # gem 'vcr', '~> 4.0'
+  gem 'vcr'
+
+  # Tests performance
+  gem 'test-prof'
+
+  gem 'database_cleaner', '~> 1.7.0'
```
Comment (icapuccio) [inline Gemfile]: We could try to remove this gem as we discussed

File: spec/support/spec_helper.rb [hunk 13]
```diff
@@ -29,13 +29,19 @@
 
   Faker::Config.random = Random.new(config.seed)
 
-  config.before(:all) do
-    FactoryBot.reload
+  config.before(:suite) do
+    DatabaseCleaner.strategy = :transaction
```
Comment (icapuccio) [inline spec/support/spec_helper.rb]: We should delete all this config related to database cleaner as well

**Review by icapuccio — APPROVED** [review]

**Review by marciondg — APPROVED** [review]
Go chaca!! :red_circle: :white_circle: :black_circle:

**Pull request merged**